src/ImageSharp/Formats/Jpeg/Components/FastFloatingPointDCT.Intrinsic.cs

```diff
@@ -0,0 +1,243 @@
+        /// </remarks>
+        /// <param name="blockRef">Input reference to the first </param>
+        public static void FDCT8x4_sse(ref Vector128<float> blockRef)
+        {
+            DebugGuard.IsTrue(Sse.IsSupported, "Sse support is required to execute this operation.");
+
+            Vector128<float> tmp0 = Sse.Add(Unsafe.Add(ref blockRef, 0), Unsafe.Add(ref blockRef, 14));
+            Vector128<float> tmp7 = Sse.Subtract(Unsafe.Add(ref blockRef, 0), Unsafe.Add(ref blockRef, 14));
+            Vector128<float> tmp1 = Sse.Add(Unsafe.Add(ref blockRef, 2), Unsafe.Add(ref blockRef, 12));
+            Vector128<float> tmp6 = Sse.Subtract(Unsafe.Add(ref blockRef, 2), Unsafe.Add(ref blockRef, 12));
+            Vector128<float> tmp2 = Sse.Add(Unsafe.Add(ref blockRef, 4), Unsafe.Add(ref blockRef, 10));
+            Vector128<float> tmp5 = Sse.Subtract(Unsafe.Add(ref blockRef, 4), Unsafe.Add(ref blockRef, 10));
+            Vector128<float> tmp3 = Sse.Add(Unsafe.Add(ref blockRef, 6), Unsafe.Add(ref blockRef, 8));
+            Vector128<float> tmp4 = Sse.Subtract(Unsafe.Add(ref blockRef, 6), Unsafe.Add(ref blockRef, 8));
+
+            // Even part
+            Vector128<float> tmp10 = Sse.Add(tmp0, tmp3);
+            Vector128<float> tmp13 = Sse.Subtract(tmp0, tmp3);
+            Vector128<float> tmp11 = Sse.Add(tmp1, tmp2);
+            Vector128<float> tmp12 = Sse.Subtract(tmp1, tmp2);
+
+            Unsafe.Add(ref blockRef, 0) = Sse.Add(tmp10, tmp11);
+            Unsafe.Add(ref blockRef, 8) = Sse.Subtract(tmp10, tmp11);
+
+            Vector128<float> z1 = Sse.Multiply(Sse.Add(tmp12, tmp13), mm128_F_0_7071);
+            Unsafe.Add(ref blockRef, 4) = Sse.Add(tmp13, z1);
+            Unsafe.Add(ref blockRef, 12) = Sse.Subtract(tmp13, z1);
+
+            // Odd part
+            tmp10 = Sse.Add(tmp4, tmp5);
+            tmp11 = Sse.Add(tmp5, tmp6);
+            tmp12 = Sse.Add(tmp6, tmp7);
+
+            Vector128<float> z5 = Sse.Multiply(Sse.Subtract(tmp10, tmp12), mm128_F_0_3826);
+            Vector128<float> z2 = Sse.Add(Sse.Multiply(mm128_F_0_5411, tmp10), z5);
+            Vector128<float> z4 = Sse.Add(Sse.Multiply(mm128_F_1_3065, tmp12), z5);
+            Vector128<float> z3 = Sse.Multiply(tmp11, mm128_F_0_7071);
+
+            Vector128<float> z11 = Sse.Add(tmp7, z3);
+            Vector128<float> z13 = Sse.Subtract(tmp7, z3);
+
+            Unsafe.Add(ref blockRef, 10) = Sse.Add(z13, z2);
+            Unsafe.Add(ref blockRef, 6) = Sse.Subtract(z13, z2);
+            Unsafe.Add(ref blockRef, 2) = Sse.Add(z11, z4);
+            Unsafe.Add(ref blockRef, 14) = Sse.Subtract(z11, z4);
+        }
+
+        /// <summary>
+        /// Apply 1D floating point FDCT inplace using AVX operations on 8x8 matrix.
+        /// </summary>
+        /// <remarks>
+        /// Requires Avx support.
+        /// </remarks>
+        /// <param name="block">Input matrix.</param>
+        public static void FDCT8x8_avx(ref Block8x8F block)
```

Comment:
Is this (and the AVX IDCT) ported from libjpeg-turbo right? If yes, links to original implementations would be nice.

Reply:
FDCT was converted from scalar libjpeg-turbo implementation, I as recall libjpeg-turbo has no floating point avx implementation, only sse one. Plus it's x86 asm only and I'm too scared to read that. So both sse and avx implementations here are my port of scalar libjpeg-turbo implementation to sse/avx.

Reply:
I see, great job!

Reply:
> it's x86 asm only and I'm too scared to read that

Don't be afraid 😄 

The reason there is no AVX FDCT implementation in libjpeg-turbo is because the integer version is so much faster it made float AVX pointless.  The AVX2 integer FDCT implementation is really quite simple, and because it keeps all the values 16-bit all the way through, all steps are faster.

Before that, the SSE2 version of the integer FDCT was quite complicated (and isn't really any faster than float), but that's mostly because it pre-dates (or doesn't assume the presence of) SSSE3 and SSE4.1 instructions that greatly simplify things.

If you want to give a port of the AVX2 asm a go, I'd be happy to answer any questions.